KAFKA-3856: Cleanup Kafka Stream builder API (KIP-120)

mjsax opened the pull request.

mjsax commented:
@guozhangwang @dguy @enothereska @miguno 

This is a preview on the indented changes. Not everything is nice right now, but I want to give a first impression on my ideas about refactoring. If we agree on the way we want to do it, I can prepare a KIP for this change.

dguy commented:
@mjsax it would be useful to have an idea of what the overall plan is. It looks like `InternalTopologyBuilder` is basically what `TopologyBuilder` is currently. Then the new `TopologyBuilder` just delegates and only provides an api that is sensible to the PAPI. So far so good.
I'm not entirely sure what value `InternalKStreamBuilder` is going to provide? Why not just have `KStreamBuilder` depend on `InternalTopologyBuilder`? I think eventually we need all of the information on the topology in one place, i.e., `InternalTopologyBuilder`, such that it can be used by `StreamThread` to build the `ProcessorTopology`.

mjsax commented:
You are right about `InternalTopologyBuilder` and `TopologyBuilder`.

Currently, `KStreamBuilder` extends `TopologyBuilder` but IMHO this abstraction is not the best one because `KStreamBuilder` offers a completely new way (ie, DSL) to specify a topology and thus, it should only *use* `TopologyBuilder` but not extend it -- hence, `KStreamBuilder`now has a member variable of type `TopologyBuilder`. (There is not need to have low level PAPI methods in `KStreamBuilder`.)

Furthermore, currently `TopologyBuilder` offers some helper methods that are DSL specific and should not be part of `TopologyBuilder`. Thus, I moved them from `(Internal)TopologyBuilder` into a new internal help class `InternalKStreamBuilder`.

There a still some abstractions that are messy -- the final goal would be to have a clear separation of `TopologyBuilder` and the DSL -- if we achieve this, we can also remove `InternalKStreamBuilder` and put its helper methods as private into `KStreamBuilder` -- right now, the abstractions are still leaky in this PR with this regard.

I agree, that in final refactoring, runtime should be have any idea if a topology was build via `TopologyBuilder` or `KStreamBuilder` -- this is not the case right now and not completely fixed by this PR yet -- it's still WIP.

guozhangwang commented:
The general solution looks good to me. I am fine with `InternalTopologyBuilder` and `InternalKStreamBuilder`.

> I agree, that in final refactoring, runtime should be have any idea if a topology was build via TopologyBuilder or KStreamBuilder -- this is not the case right now and not completely fixed by this PR yet -- it's still WIP.

You mean "runtime should not have any idea"?

> There a still some abstractions that are messy -- the final goal would be to have a clear separation of TopologyBuilder and the DSL

One thing worth considering though, is that today we have seen some users using `addProcessor` and `KStream/KTable` interleavingly with `KStreamBuilder`. Honestly I think this is not a good pattern but would like to have us think what would potentially be restricted if we do not let `KStreamBuilder` to extend from `TopologyBuilder` anymore.

mjsax commented:
> You mean "runtime should not have any idea"?
Yes.

> One thing worth considering though, is that today we have seen some users using addProcessor and KStream/KTable interleavingly with KStreamBuilder. Honestly I think this is not a good pattern but would like to have us think what would potentially be restricted if we do not let KStreamBuilder to extend from TopologyBuilder anymore.

Agreed. But there should not be any limitations as we offer `KStream#transform()` that should offer all flexibility `addProcessor` offers, too. (If I don't miss anything.)

mjsax committed:
KAFKA-3856: Cleanup Kafka Streams builder API
 - add TopologyDescription (as non-public class)
 - add non-public TopologyBuilder.describe()

mjsax commented:
Call for review @enothereska @dguy @guozhangwang 

This is part of KIP-120. I try try to break it down to multiple smaller PRs. This PR does not yet add any new public API (in case we get KIP-120 not finished for `0.11`), thus we can merge it as a preparation PR.

Note, that I change the API of `TopologyDescription` slightly -- instead of `List` I use `Set` -- I did update the KIP accordingly, too. Do we need to send a note to dev list about this update?

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1378,4 +1374,168 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+        return false;
+    }
+
+    TopologyDescription describe() {
```
Should this be public?

mjsax replied:
No. We will have `public Topology#describe()` -- `TopologyBuilder` will get deprecated anyway and this method will "move" to `InternalTopologyBuilder` later (than it will get public but in package `internal`).

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1378,4 +1374,168 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+            final NodeFactory nodeFactory = nodeFactories.get(nodeName);
+            final TopologyDescription.Node node;
+
+            if (nodeFactory instanceof SourceNodeFactory) {
```
Rather than doing `if(nodeFactory instanceof blah)` could we add an abstract method `ToplogoyDescription.Node describe(...)` to `NodeFactory` and then implement it in the subclasses. My skin crawls when i see `instanceof` :-)

mjsax replied:
Ack.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1378,4 +1374,168 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+
+        if (sourcePattern != null) {
+            return sourcePattern.pattern();
+        } else {
```
nit: personal preference so feel free to ignore. But i'd ditch the `else` in this case. I don't think it adds anything

mjsax replied:
I personally think, it reads better (at least my mind has an easier time parsing this).

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1378,4 +1374,168 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+                predecessorNames = ((SinkNodeFactory) nodeFactory).parents;
+
+            } else {
+                assert nodeFactory instanceof SourceNodeFactory;
```
why bother with this? `NodeFactory` is private in `TopologyBuilder` so it is not going to be anything else. Plus it wouldn't matter if we used the interface rather than `instanceof`

mjsax replied:
> it is not going to be anything else
That's true atm. But why do we write a test, the code work already :) (same miss leading argument) -> it's a guard against adding new "types" of `NodeFactory` -- it forces you to update this part of the code, too (otherwise you might miss it) And yes, it's private, thus it's an assert and no exception.

Not sure who the interface change resolve this issue? A `SourceNodeFactory` will have have `#parents` and thus, we will need to cast here... Or do I miss something?

dguy replied:
if you added a method to `NodeFactory` `String[] parents()` then in `SourceNodeFactory` you implement it by returning `new String[0]`
This code becomes 
```
String[] predecessorNames = nodeFactory.parents()
for(final String predecessorName : predecessorNames) {
    ...
}
```

mjsax replied:
ok.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1378,4 +1374,168 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+            final NodeFactory nodeFactory = nodeFactories.get(node.name());
+
+            final String[] predecessorNames;
+            if (nodeFactory instanceof ProcessorNodeFactory) {
```
same as before. Would be much nicer to add a method on the abstract class rather than using instanceof

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+    public final Set<GlobalStore> globalStores = new HashSet<>();
+
+    /**
+     * A connect sub-graph of a {@link Topology}.
```
'A connected ...' ?

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+     */
+    public final static class Source implements Node {
+        /** The name of this source node. */
+        final String name;
```
why not private?

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+         * The topics name this source node is reading from.
+         * This can be comma separated list of topic names or pattern (as String).
+         */
+        public final String topics;
```
ditto

mjsax replied:
Could make it private, but than I need to add a getter... Does this improve anything?

dguy replied:
This will eventually be a public API - right? I'd add a getter for it if that is the case

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+
+    public final static class Processor implements Node {
+        /** The name of this processor node. */
+        final String name;
```
Same here and everywhere else. I don't see why the members aren't private?

dguy commented on streams/src/test/java/org/apache/kafka/streams/processor/TopologyTest.java:
```diff
@@ -0,0 +1,412 @@
+    // TODO remove -- only for testing during coding
+    @After
+    public void after() {
+        System.out.println(topology.describe());
```
Should we just remove this so it doesn't get committed?

mjsax replied:
Ack.

dguy reviewed (commented):
Nice set of tests added! A few comments left.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+
+        @Override
+        public int hashCode() {
+            int result = id;
```
Why not simply

```java
return 31 * id + nodes.hashCode();
```
?

mjsax replied:
Autogenerated code... :)

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+
+        @Override
+        public int hashCode() {
+            int result = source.hashCode();
```
See above.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+     * A node of a topology. Can be a source, sink, or processor node.
+     */
+    public interface Node {
+        /** The name of the node. */
```
What's the contract for:

- null values?
- empty string?

mjsax replied:
According to `TopologyBuilder` empty string would be ok -- `null` is not allowed. But do we need to document this here, too? Or even enforce in the constructor (that is not public) ?

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+         * The predecessors of this node within a sub-topology.
+         * Note, sources do not have any predecessors.
+         */
+        Set<Node> predecessors();
```
What's the contract for null values?  Ok to return null?

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+            }
+
+            final Source source = (Source) o;
+            // oit successor to avoid infinite loops
```
"oit"?

mjsax replied:
typo: omit :)

mjsax committed:
Github comments

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+            }
+
+            final Processor processor = (Processor) o;
+            // oit successor to avoid infinite loops
```
"oit"?

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+        public int hashCode() {
+            int result = name.hashCode();
+            result = 31 * result + stores.hashCode();
+            result = 31 * result + predecessors.hashCode();
```
This looks odd.  Why the same prime?

Also, as a general comment for the various `hashCode()` implementations in this PR, why don't we use [Objects#hash()](https://docs.oracle.com/javase/7/docs/api/java/util/Objects.html#hash(java.lang.Object...)) instead of manually doing this?

mjsax replied:
findbugs -> it's bad to overwrite`equals` but not `hashCode`

mjsax replied:
about same prime number -> auto generated code
I have personally no experience with writing a good hashCode function. If you have any (more concrete) advice, I am happy to incorporate it :)

miguno replied:
Findbugs: What I meant was:  Instead of manually doing the above computations inside `hashCode()`, we can simply rely on `Objects#hash()` (was introduced in Java 7) to do these computations for us.

mjsax replied:
Ack

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+        public int hashCode() {
+            int result = name.hashCode();
+            result = 31 * result + topic.hashCode();
+            result = 31 * result + predecessors.hashCode();
```
Same as above.

miguno commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,412 @@
+
+    @Override
+    public int hashCode() {
+        int result = subtopologies.hashCode();
```
See above.

miguno commented:
No unit tests for Subtopology, GlobalStore, Source, Processor, Sink?

mjsax commented:
> No unit tests for Subtopology, GlobalStore, Source, Processor, Sink?

There is nothing to test IMHO -- it's just getter and setters... those classes are more like plain "structs".

mjsax commented:
Updated this.

mjsax commented:
Builds fails for Java8 -- can reproduce locally -- looking into it...

mjsax committed:
fixed failing test and rebased

mjsax commented:
Updated this. @guozhangwang @dguy @enothereska @bbejeck

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5718/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5732/
Test PASSed (JDK 7 and Scala 2.11).

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1558,4 +1578,98 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+            }
+        }
+
+        description.subtopologies().add(new TopologyDescription.Subtopology(
```
Maybe add an `addSubtopology` method? I don't really like the idea of a) exposing, and b) mutating an internal collection.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1558,4 +1578,98 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+                    it.remove(); // remove sourceNode from group
+                    final String processorNode = nodes.iterator().next(); // get remaining processorNode
+
+                    description.globalStores().add(new TopologyDescription.GlobalStore(
```
same as above comment

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,480 @@
+         * @return set of all nodes within the sub-topology
+         */
+        public Set<Node> nodes() {
+            return nodes;
```
Should we return an unmodifiable set?

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,480 @@
+            }
+
+            final Subtopology that = (Subtopology) o;
+            return id == that.id
```
Is it enough here to just compare `id` ? I.e, it is unique

mjsax replied:
Yes. It's the same as `nodeGroup` ID as used internally by `TopologyBuilder`

dguy replied:
what i meant was i don't think we also need `nodes.equals(that.nodes)` but maybe we probably do as the id is not unique across topologies

mjsax replied:
Exactly.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,480 @@
+
+            final Source source = (Source) o;
+            // omit successor to avoid infinite loops
+            return name.equals(source.name)
```
Can we just compare `name` here? It is unique, right?

mjsax replied:
Yes. All node names (ie, (`ProcessorName`s) are unique. Maybe we should make `Node` an abstract class and push the `name` there?

mjsax replied:
But they are unique with a topology. If you have two topologies, both could use the same name but different input topics -- thus, both would not be equal. Thus, I think we should use `source.topics` here -- same below. Let me know what you think.

dguy replied:
Makes sense, I guess people could use it like that.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,480 @@
+
+            final Processor processor = (Processor) o;
+            // omit successor to avoid infinite loops
+            return name.equals(processor.name)
```
same here?

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,480 @@
+                return false;
+            }
+
+            final Sink sink = (Sink) o;
```
As above

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,480 @@
+     * @return set of all sub-topologies
+     */
+    public Set<Subtopology> subtopologies() {
+        return subtopologies;
```
return unmodifiable Set ? Also next method

dguy reviewed (commented):
@mjsax i haven't finished reviewing, but is this PR just for the `TopologyDescription` part and not the rest of the cleanup?

mjsax committed:
Damian's comments

mjsax commented:
@dguy I updated the PR. And yes, it's only one part for KIP-120. Small PRs FTW :)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5828/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5814/
Test PASSed (JDK 8 and Scala 2.12).

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -172,6 +174,11 @@ public void addStateStore(String stateStoreName) {
         }
+
+        @Override
+        TopologyDescription.Processor describe() {
```
Any reason why we don't just stick with return `AbstractNode` from these overrides? I don't see anywhere that we need it to be a `Processor` etc

mjsax replied:
It's just to have stronger type safety. We know, that we only return `Processor` but nothing else. I don't see any disadvantage in having a strong type.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1558,4 +1578,98 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+                                     final Integer subtopologyId,
+                                     final Set<String> nodeNames) {
+
+        final HashMap<String, TopologyDescription.Node> nodesByName = new HashMap<>();
```
if we make this `<String, TopologyDescription.AbstractNode>` then can we do away with the casts below?
I think we know that they will always be `AbstractNode`

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,476 @@
+import java.util.Set;
+
+/**
+ * A meta representation of a {@link Topology topology}.
```
Should we hold off on putting the links to `Topology` in until the class exists?

mjsax replied:
I think, doing a cleanup later on makes it more complex and error prone. And it's just in `trunk` for not long as the next PR is in the making already...

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,476 @@
+         * The source node reading from a "global" topic.
+         * @return the "global" source node
+         */
+        public Source source() {
```
I wonder if we should return an interface here, i.e., once that only has the read/get methods? `Source` has `addPredecessor` and `addSuccessor`, which i don't think is correct for the user facing interfaces.
Same for the next method.

mjsax replied:
Both `addPredecessor` and `addSuccessor` are package private and cannot be accessed by the user. So should be fine (should also not show up in the JavaDocs) IMHO. But we can change if you think it's better design (but what do we gain?)

dguy reviewed (commented):
I've left a few more comments. Can we get a second review from @enothereska @bbejeck or @guozhangwang ?

mjsax committed:
Damian's comments

mjsax commented:
Updated to use `AbstractNode`

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5865/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5850/
Test PASSed (JDK 8 and Scala 2.12).

dguy reviewed (approved):
LGTM, but would still like a second review from  @enothereska @bbejeck or @guozhangwang

mjsax commented:
Call for review @enothereska @guozhangwang

bbejeck reviewed (approved):
LGTM

dguy commented:
retest this please

dguy commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6152/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6168/
Test PASSed (JDK 7 and Scala 2.11).

dguy commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6182/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6198/
Test PASSed (JDK 7 and Scala 2.11).

dguy commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6192/
Test FAILed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6208/
Test FAILed (JDK 7 and Scala 2.11).

dguy commented:
test look unrelated, but i'd like to see a clean build before merging

dguy commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6197/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6213/
Test PASSed (JDK 7 and Scala 2.11).

dguy commented:
Thanks @mjsax, Merged to trunk

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -144,23 +144,25 @@
         ProcessorNodeFactory(String name, String[] parents, ProcessorSupplier<?, ?> supplier) {
-            super(name);
-            this.parents = parents.clone();
+            super(name, parents.clone());
```
The parameters can be `final`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -187,7 +194,7 @@ private SourceNodeFactory(final String name,
                                   final Deserializer<?> keyDeserializer,
                                   final Deserializer<?> valDeserializer) {
-            super(name);
+            super(name, new String[0]);
```
nit: we can define a static

```
private static final String[] NO_PARENTS = {};
```

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -236,18 +243,30 @@ public ProcessorNode build() {
         }
+
+        @Override
+        TopologyDescription.Source describe() {
```
I'm wondering why we keep the description construction of sink and process nodes in `TopologyDescription`, while for source nodes it is kept here? Is this inconsistency intentional?

mjsax replied:
It's all the same. There is no inconsistency IMHO. All three constructors are just setting members and we call a `new` for all three in the corresponding `NodeFactory` -- the only different is, that for source we need to distinguish between pattern vs topic subscription and need a new lines of code to set up the constructor argument (this step is not required for processors and sinks)

guozhangwang replied:
You are right. NVM.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,476 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.processor;
```
I think the package is incorrect: in KIP it is described as in `o.a.k.streams`.

mjsax replied:
You are right that KIP and code don't align. But I am not sure if KIP or code is "correct". As `Topology` is contained in `o.a.k.streams.processor` it seems correct to put `TopologyDescription` into it, too. Thoughts?

\cc @dguy @enothereska @bbejeck

guozhangwang replied:
Hmm... in KIP it is described as `org.apache.kafka.streams.Topology` so I think the newly adding `Topology` will also be in the `o.a.k.streams`, which looks the right place to me than `o.a.k.streams.processor`.

mjsax replied:
Good point. KIP seems to be correct. I guess because `TopologyBuilder` is in `.processor` I put `TopologyDescription` into the wrong package. Will fix this in #3536

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,476 @@
+ * sub-topology {@link Topology#addSink(String, String, String...) writes} into a topic and another sub-topology
+ * {@link Topology#addSource(String, String...) reads} from the same topic.
+ * <p>
+ * For {@link KafkaStreams#start() execution} sub-topologies are translated into {@link StreamTask tasks}.
```
nit: "translated" may read a bit misleading to users. How about sth. like "when ... is called, different sub-topologies will be constructed and executed as independent {link StreamTask tasks}".

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,476 @@
+         * Will never be {@code null}.
+         * @return set of all predecessors
+         */
+        Set<Node> predecessors();
```
I like `predecessors`! Could we change the term of `parents` in `TopologyBuilder` as well?

mjsax replied:
Ack.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,476 @@
+        private String nodesAsString() {
+            final StringBuilder sb = new StringBuilder();
+            for (final Node node : nodes) {
+                sb.append("    ");
```
nit: `\t`?

mjsax replied:
`\t` is usually rendered as 8 spaces what is a little bit wide for my taste :) Quite subjective though... I did this intentionally, but I am also open to change it.

guozhangwang replied:
Makes sense let's leave it as is :) Just checking if you have thought about this already.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyDescription.java:
```diff
@@ -0,0 +1,476 @@
+            sb.append("  none\n");
+        } else {
+            for (final Subtopology st : subtopologies) {
+                sb.append("  ");
```
One meta comment: if we want to use `\t` or `    ` as indents then we need to pass in through the sub-calls since otherwise only the first line will be indented.

mjsax replied:
The indention is basically hard coded and this renders fine. We don't render the structure of sub-topologies but only list all nodes within it. I did print `GlobalKTableIntegartionTest#shouldKStreamGlobalKTableLeftJoin`:
```
Sub-topologies: 
  Sub-topology: 2
    Processor: KTABLE-SOURCE-0000000004(stores: [table]) -->  <-- KSTREAM-SOURCE-0000000003
    Source: KSTREAM-SOURCE-0000000003(topics: input-table-1) --> KTABLE-SOURCE-0000000004
  Sub-topology: 1
    Processor: KSTREAM-LEFTJOIN-0000000005(stores: []) --> KSTREAM-FOREACH-0000000006 <-- KSTREAM-SOURCE-0000000002
    Processor: KSTREAM-FOREACH-0000000006(stores: []) -->  <-- KSTREAM-LEFTJOIN-0000000005
    Source: KSTREAM-SOURCE-0000000002(topics: input-stream-1) --> KSTREAM-LEFTJOIN-0000000005
Global Stores:
  GlobalStore: KSTREAM-SOURCE-0000000000(topic: globalOne-1) -> KTABLE-SOURCE-0000000001(store: globalStore)
```

The "arrows" indicate predecessor and successors for each node. It's not quite pretty and I am happy to improve this if you have good idea. However, the main purpose of this API is not to just print the topology to stdout but you use if for own rendering. Also, anybody can to a stdout pretty print too by implementing their own traversal strategy that produces a `String`.

guozhangwang replied:
Ack. One other nit comment: when `predecessors` or `successors` are empty could we print a `none` like we did for sub-topologies? Right now Strings like `-->  <--` looks a bit confusing.

mjsax replied:
Ack.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1558,4 +1578,98 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+        return false;
+    }
+
+    private void describeSubtopology(final TopologyDescription description,
```
Another meta comment: we now have two similar code path, one for building the topology for execution, and one for building the topology description for printing. Could we consider consolidate these two?

Just a wild thought, do not need to be done in this PR / KIP.

mjsax replied:
Maybe... If you think it's important, we can create a Jira to track it.

guozhangwang replied:
Maybe we can create a JIRA for tracking, but it's not that important for now. Since the common parts that can be consolidated may be not much: in the actual topology building process we need to set the internal topic names, set copartition topics etc which are not needed for the topology description building at all. 

What I was originally thinking is the the topology building process may be extending from the topology description building process with its additional functionalities like I mentioned above, but I am all hand-wavy on the devil details now.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java:
```diff
@@ -1558,4 +1578,98 @@ public synchronized void updateSubscriptions(final SubscriptionUpdates subscript
+
+                if (isGlobalSource(node)) {
+                    // we found a GlobalStore node group; those contain exactly two node: {sourceNode,processorNode}
+                    it.remove(); // remove sourceNode from group
```
nit: this comment after the code seems unnecessary to me.

guozhangwang reviewed (commented):
@mjsax Sorry for the late review. I have a few comments which could be addressed in https://github.com/apache/kafka/pull/3536, or another follow-up PR.
